### ISSUE TITLE
Custom class Nodes break when changing inheritance

### ISSUE THREAD

**Issue by Aliviessa** [godotengine/godot#81222]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Windows 10

### Issue description

I have created a custom class (let's call it CustomNode2D)  inheriting Node2D and subsequently created a ton of scenes with that node type. Afterwards i modified the class to inherit Area2D. This leads to the helpful error message 'Script inherits from native type 'Area2D', so it can't be assigned to an object of type 'Node2D'

This can be fixed by opening the scene and "changing" the node type to CustomNode2D. This feels unnecessary, because when hovering the node it correctly shows its Type as CustomNode2D. Perhaps there should be a way to update all scenes with nodes based on a modified class to avoid having to manually click through all these scenes.

### Steps to reproduce

- Create class inheriting Node2D

```
#CustomNode2D.gd
class_name CustomNode2D
extends Node2D
```

- Create scene with root node based on CustomNode2D
- Modify CustomNode2D to inherit another class
```
#CustomNode2D.gd
class_name CustomNode2D
extends Area2D
```
- Run previously created scene 

### Minimal reproduction project

[New Game Project.zip](https://github.com/godotengine/godot/files/12497612/New.Game.Project.zip)

**Comment by mieldepoche** [comment]
If you create a node from your `Node2D`-extending script, it will be a `Node2D`.
You can't put an `Area2D`-extending script on a `Node2D`, hence the error.

There is a bug regarding the icon used for the node the script is attached to though, it shows the icon of the type the script extends.

**Comment by AThousandShips** [comment]
This is indeed not a bug, you cannot attach a script of class `A` to a node of class `B` unless `A` is `B` or a class deriving from `B`, it just isn't possible as you can't apply the behaviour of `A`

**Comment by astrowild99** [comment]
While trying to replicate the problem with an addon implementing the `CustomNode2D`, nothing actually stops the user from changing the type.  
Maybe it would be great to have a `Node Configuration Warning` like this one?
![Screenshot from 2023-09-03 19-12-16](https://github.com/godotengine/godot/assets/20246421/a78bbac1-5602-4c6c-a12b-05bbbd447be1)

**Comment by Aliviessa** [comment]
> If you create a node from your `Node2D`-extending script, it will be a `Node2D`. You can't put an `Area2D`-extending script on a `Node2D`, hence the error.

I understand that. However when inspecting the scene it will show the following:
![image](https://github.com/godotengine/godot/assets/125681566/71a23c68-082f-4c3d-97a7-175f270c0364)

This looks correct no? It however will show the previously mentioned error until i change type to the same type that is already shown.

**Comment by AThousandShips** [comment]
That is correct, the explicit type is that of the script, if a `Node2D` has a script attached that is a `Node2D` extended type called `MyNode2D` then the type of the node is `MyNode2D`

**Comment by Aliviessa** [comment]
> That is correct, the explicit type is that of the script, if a `Node2D` has a script attached that is a `Node2D` extended type called `MyNode2D` then the type of the node is `MyNode2D`

Ok so if i understand correctly there are 2 types associated with a node: the node type, that is derived from me setting up a node via "add child node" or "change type" and the script type, that is derived from the attached script?

**Comment by mieldepoche** [comment]
> there are 2 types associated with a node

kiiinda? (see #81290)
though this doesn't generally matter too much, since the script you attach to a node usually superseedes or directly extends the node's type, so the script type doesn't conflict with the type of the node it's attached to (in my experience at least)

**Comment by bjornmp** [comment]
This is how it looks like in my type enforcement branch. That branch is actually about general custom type consistencies, but I also added this check because It was at the same spot where I was doing changes anyway.

![image](https://github.com/godotengine/godot/assets/8940144/35d06732-92ca-4e81-a044-15217a45f96b)

**Comment by h-sigma** [comment]
Imagine this scenario where:
`Node2D -> EntityBase` is a parent -> sub class relation. If, officially speaking, we can assign  the `EntityBase` script on any node that inherits `Node2D` even if does not have `EntityBase` in its inheritance chain, that would allow a lot of code-reuse especially when it comes to built-in types whose inheritance chain you can't extend right now. As an example, for a small framework I'm making where entities may be CharacterController2D's or plain old Node2Ds.
